Refactor remote job engine to Engine/Connection API

pauljxtan opened the pull request:
**Context:**

Strawberry Fields provides engines for computing quantum programs against local and remote (currently on the `demo` branch) backends. The purpose of this PR is to refactor and simplify the existing functionality on `demo` to an API structured around two core objects, an `Engine` and a `Connection`, to facilitate envisioned job execution workflows.

**Description of the Change:**

- Refactor the existing `StarshipEngine` and `APIClient` to define interfaces at two distinct levels of abstraction: a higher-level `StarshipEngine` that provides the minimal methods needed to run a job (more intended for users), and a lower-level `Connection` that provides more flexibility around job management (more intended for developers or automated testing environments).
- The `StarshipEngine` allows remote jobs to be run synchronously (via `engine.run()`) or asynchronously (via `engine.run_async()`). In the former mode, the engine blocks until the job completes, then returns the final result. In the latter mode, the engine immediately returns a job object to be manually queried and refreshed by the user.
- The `Connection` exposes additional remote platform operations, such as cancelling a job in progress or directly querying its status or result. The `StarshipEngine` serves as a high-level wrapper around this object, but the `Connection` can also be instantiated and used directly, independent of the engine.
- The `starship` CLI is refactored to use the new classes, and also simplified to a more minimal interface.

N.B. The changed files are auto-formatted with `black` (default settings), which produces a number of purely cosmetic diffs.

**Benefits:**

- Provide simplified abstractions to ease various job execution workflows from both a user and developer perspective.
- Job results are retrieved from the remote platform in `numpy` format (with `dtype=np.int8`) instead of JSON, reducing bandwidth usage and conversion overhead.

**Possible Drawbacks:**

- This will likely break downstream applications currently submitting jobs via the demo branch, so some coordination will be needed there.

**Related GitHub Issues:**

N/A

pauljxtan commented on tests/frontend/test_engine.py:
```diff
@@ -166,209 +174,297 @@ def inspect():
-        for the hardware backend, is locked, is added to self.run_progs, that it is run and that
-        a Result object is returned populated with the result samples.
-        """
+        assert np.array_equal(result.samples.T, result_samples)
```
This probably deserves some explanation. As the `Result` class is currently implemented, `Result.samples` has dimensions `(modes, shots)`; on the other hand, job results received from the remote backend have the shape `(shots, modes)`, so we transpose the samples here. Modifying `Result.samples` to return `(shots, modes)` appears to break various tests elsewhere in SF, so I opted to leave it unchanged. I also believe that the issue of sample dimensions is being (or has been) addressed on `master`, so it may be better to revisit this in another PR down the line.

antalszava replied:
Thanks for the explanation! One thing to note is that there is an upcoming PR to be merged on this: #300

thisac replied:
Just mentioning that #300 is already merged (earlier today). It unified the sample returns for all simulator engines so that they return NumPy arrays with dimension `(shots, modes)`.

pauljxtan replied:
Cool, thanks for the heads-up guys. I'm assuming we will have another substantial PR at some point to line up `demo` with `master`, so that may be a good time to get all the dimensions agreeing?

pauljxtan commented on strawberryfields/engine.py:
```diff
@@ -531,161 +544,541 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
+            )
+        if connection is None:
+            # TODO use `load_config` when implemented
+            config = DEFAULT_CONFIG["api"]
```
This line should be updated to use `load_config()` after https://github.com/XanaduAI/strawberryfields/pull/298 is merged.

antalszava replied:
It got merged just now :)

pauljxtan replied:
Great, I've updated the relevant logic to use the new configuration functions!

pauljxtan commented on strawberryfields/engine.py:
```diff
@@ -531,161 +544,541 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
+                # > after
+            # ]
+        # raise RequestFailedError(self._format_error_message(response))
+        raise NotImplementedError("This feature is not yet implemented")
```
I've left the above commented for context, but I think we probably want to remove it before merging.

pauljxtan replied:
On further thought, I wonder if it's worth having this method stub here at all, or if we should just remove it and put it back in a future PR. Thoughts @antalszava @josh146?

antalszava commented on strawberryfields/engine.py:
```diff
@@ -531,161 +544,541 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
+    PATCH = "patch"
+
+
+class Connection:
```
Hmm, this file seems to be getting quite big, maybe it would be worth having the `Connection` class, `Job` and related `Error` classes in a separate one called `connection.py`? :thinking: Or to split them is some other meaningful way :blush:

pauljxtan replied:
Great idea! Especially as the `Connection`/`Job` interface is intended to be usable independent of the `StarshipEngine`, and the module boundary would help to enforce and make explicit that separation.

One problem I see though is that creates a circular dependency due to the `engine.Result` class, which would to be imported in `connection.py`---and `engine.py` would need to import `Connection` from `connection.py`, etc.

A solution to that might be to move `Result` into a separate module. What do you think?

antalszava replied:
Sounds good to me :)

josh146 replied:
:+1: I commented the exact same thing last night, but couldn't submit my review because github buttons stopped working 😆 But yes, I agree. Options include:

* Package: `api/connection.py`, `api/job,py`, etc.

* A module `connection.py` that contains all classes

I tend to prefer option (1), an API package, as it collects all objects as the same level of abstraction.

pauljxtan replied:
I've moved the relevant code into the following modules: `api/connection.py`, `api/job.py`, and `api/result.py`.

antalszava commented on strawberryfields/engine.py:
```diff
@@ -531,161 +544,541 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
+        token (str): the API authentication token
+        host (str): the hostname of the remote platform
+        port (int): the port to connect to on the remote host
+        use_ssl (bool): whether to use SSL for the connection
```
In the configuration, a `debug` boolean is also specified, I'm missing more details on where it should be incorporated. Maybe it would be intended here as well? :thinking:  @josh146 could you help out with that one?

pauljxtan replied:
Ah, yes - the former `APIClient` implemented a `debug` mode that simply collected responses and errors in list attributes on the client object. I didn't see a particular use case for that, so left it out for simplicity; but it should not be a very big change to add that in if deemed necessary.

antalszava replied:
I see! Let's see if there would be a use case. Currently, it is included in #298, but that could also be adjusted accordingly. :)

josh146 replied:
Perhaps, rather than debug, a better options would be `verbose=True`. In verbose mode, SF simply prints out details of the job submission to the terminal as it occurs.

@antalszava, might make sense to remove `debug` from #298 for now

pauljxtan replied:
The `verbose` option has been added!

antalszava commented on strawberryfields/engine.py:
```diff
@@ -531,161 +544,541 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
+        self, method: RequestMethod, path: str, headers: Dict[str, str] = None, **kwargs
+    ) -> requests.Response:
+        headers = {} if headers is None else headers
+        return getattr(requests, method.value)(
```
It might be worth having a small auxiliary naming here for readabilitiy such as
```python
request_method = getattr(requests, method.value)
return request_method(urljoin(self.base_url, path),
            headers={"Authorization": self.token, **headers},
            **kwargs
        )
```

pauljxtan replied:
Thanks for pointing that out! Updated.

antalszava replied:
Hmm, after going through `_get`, `_post`, `_patch` and `_request`, I'm wondering, is there any advantage in having `_request` and having a `RequestMethod` enum class?

Maybe the explicit values of `get`, `post`, and `patch` could simply be used in their corresponding methods to making the return statement for example for `_get`:
```python
return requests.get(urljoin(self.base_url, path), headers={"Authorization": self.token, **headers}, **kwargs)
```

But let me know if I missed something or you'd rather keep the implementation as is, for future extendability, this is more of a suggestion after parsing each of them separately. :)

pauljxtan replied:
Yes, that's a perfectly valid point :+1:

I was on the fence myself about whether that introduced additional indirection for no substantial benefit. My thinking was to avoid repeating shared elements, like the `"Authorization"` header, between requests; but now that you mention it, it seems that is the _only_ thing that is actually duplicated, so it's probably better to just be explicit and direct about it.

I will look at simplifying the code as you described. Thanks!

pauljxtan replied:
I've removed the wrapper methods for sending requests, and the `create_job`, `get_job` (etc.) methods now call the relevant `requests` method directly.

antalszava commented on strawberryfields/engine.py:
```diff
@@ -531,161 +544,542 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
+            )
+        raise RequestFailedError(
+            "Job creation failed: {}".format(self._format_error_message(response))
+        )
```
Cool! :)

antalszava commented on strawberryfields/engine.py:
```diff
@@ -531,161 +544,542 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
+    FAILED = "failed"
+
+    @property
+    def is_terminal(self) -> bool:
```
Just a minor suggestion here: would it make sense to go with the docstring and name this `is_final`? :slightly_smiling_face:

pauljxtan replied:
Good catch, updated :+1:

antalszava commented on strawberryfields/engine.py:
```diff
@@ -531,161 +544,542 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
-            )
+        response = self._patch(
+            "/jobs/{}".format(job_id), data={"status", JobStatus.CANCELLED.value}
+        )
```
Just for clarification: you are using `PATCH` here such that you can modify *only* the status, right?

pauljxtan replied:
That's right!

antalszava commented on strawberryfields/engine.py:
```diff
@@ -531,161 +544,542 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
 
-    def __init__(self, backend, polling_delay_seconds=1, **kwargs):
-        super().__init__(backend)
+        This is only defined for complete jobs, and raises an exception for any other
```
```suggestion
        This is only defined for completed jobs and raises an exception for any other
```

antalszava replied:
Oh, this suggestion regards other parts as well, such as `JobStatus.COMPLETE`. I thought it was more descriptive to have `completed`, but this is pretty subjective and let me know if you'd rather keep it as is :)

pauljxtan replied:
No, I totally agree that "completed" makes more sense in this context (as a synonym for "finished"), while "complete" sounds to me along the lines of "full" or "whole".

The reason I had it as "complete" to begin with is because that is the identifier used by the remote platform :smiley: So the value of the `JobStatus` enum would have to be `"complete"` to match up with the responses, but the enum name could certainly be `COMPLETED` (along with comments, docstrings, etc.). I will make the change, thanks!

antalszava replied:
Oh, I see! Feel free to leave it as is, if you'd rather not have any further confusions coming from this difference :)

pauljxtan replied:
I've updated all occurrences of "complete" to "completed" except for the single value in the enum member. Don't think it should lead to much confusion, but I'll add a comment to clarify,

antalszava replied:
Great, thanks! :)

antalszava commented on strawberryfields/engine.py:
```diff
@@ -531,161 +544,542 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
+            )
+        self._status = self._connection.get_job_status(self.id)
+        if self._status == JobStatus.COMPLETE:
+            self._result = self._connection.get_job_result(self.id)
```
Cool! :)

antalszava commented on strawberryfields/engine.py:
```diff
@@ -126,6 +128,8 @@ def state(self):
             BaseState: quantum state returned from program execution
         """
+        if not self._is_stateful:
+            raise AttributeError("The state is undefined for a stateless computation.")
```
Great! :)

antalszava commented on strawberryfields/engine.py:
```diff
@@ -531,161 +544,542 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
+    def _patch(
+        self, path: str, headers: Dict[str, str] = None, **kwargs
+    ) -> requests.Response:
+        return self._request(RequestMethod.PATCH, path, headers, **kwargs)
```
Maybe these could be made static, what do you think?

pauljxtan replied:
Went one better and removed them completely as we now use the `requests` functions directly :slightly_smiling_face:

antalszava commented on strawberryfields/engine.py:
```diff
@@ -531,161 +544,542 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
+                if job.status == JobStatus.COMPLETE:
+                    return job.result
+                if job.status == JobStatus.FAILED:
+                    raise JobFailedError("The computation failed; please try again.")
```
I wonder if we'll be able to provide more information here to the user on the cause of failure (e.g. timeout, etc.), but right now I cannot really think of anything else.

josh146 replied:
This is something that should be probably done via logging (i.e., `log.info(message)`, `log.warning(message)` etc).

pauljxtan replied:
For that particular case (job status becomes "failed"), I believe the most detail we can provide to the user is something along the lines of "The computation failed due to an internal system error." (There may be more specific causes, e.g. hardware error, that is only available in our internal logs.)

@josh146 Do you mean logging a failure message and not raising an exception at all?

antalszava commented on strawberryfields/engine.py:
```diff
@@ -531,161 +544,542 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
-            Result: results of the computation
+            strawberryfields.engine.Job: the created remote job
         """
+        return self._connection.create_job(self.target, program, shots)
```
This looks really neat! :)

antalszava commented on tests/frontend/test_engine.py:
```diff
@@ -166,209 +174,299 @@ def inspect():
+        result = Result([[1, 2], [3, 4]], is_stateful=False)
+
+        with pytest.raises(AttributeError):
+            _ = result.state
```
Nice!

antalszava commented on strawberryfields/engine.py:
```diff
@@ -531,161 +544,507 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
+        if self.status.is_final:
+            raise InvalidJobOperationError(
+                "A {} job cannot be refreshed".format(self.status.value)
+            )
```
I'm wondering about raising an error in this case. If the user mistakenly calls on `refresh`, then this would throw an `InvalidJobOperationError` in their script and terminate it. Instead of that, maybe information could be provided, why no refreshing can take place (e.g. the result is already contained, or the job failed, etc.) and perhaps a warning could be raised? :thinking: 

What do you think?

antalszava replied:
(For cancellation, it would seem in order to raise an error, as the user expects to take action, whereas `refresh` seems to be something more along the lines of trying to extract more information about the `Job`. But this might just be my point of view.)

josh146 replied:
Agree

pauljxtan replied:
Yeah, that makes sense. I'll replace that raised exception with a warning message.

pauljxtan replied:
That exception has been replaced with a logged warning message.

antalszava commented on tests/frontend/test_engine.py:
```diff
@@ -166,209 +175,299 @@ def inspect():
-        assert len(starship_engine.jobs) == 0
 
-    def test__get_blackbird(self, starship_engine, monkeypatch):
+def mock_response(status_code, json_body=None, binary_body=None):
```
If you would like, this could also become a `MockResponse` class and then there would be no need to use `MagicMock` (if I'm not missing something). The `__init__` could then contain the branching logic. The upside of that would be having `pytest` as the only testing framework in this file.

pauljxtan replied:
Great idea. I've replaced `mock_response` with a `MockResponse` class!

antalszava commented on tests/frontend/test_engine.py:
```diff
@@ -166,209 +175,299 @@ def inspect():
+
+    def test_cancel_job(self, connection, monkeypatch):
+        """Tests a successful job cancellation request."""
+        monkeypatch.setattr(requests, "patch", mock_return(mock_response(204, {})))
```
One thing to note here, that if I understand it correctly any PATCH request is mocked out this way, correct?

One thing that could be checked here, is that `requests.patch` receives the correct arguments. This would include creating a function similar to `mock_return`, but one, that also tracks the arguments passed, such as:
```python

arguments = []
def mock_return(return_value):
    """A helper function for defining a mock function that returns the given value for
    any arguments.
    """
    def function(*args, **kwargs):
        arguments.append(args)
        return return_value
    return function
```
or something comparable.

```python
In [2]: arguments = [] 
   ...: def mock_return(return_value): 
   ...:     """A helper function for defining a mock function that returns the g
   ...: iven value for 
   ...:     any arguments. 
   ...:     """ 
   ...:     def function(*args, **kwargs): 
   ...:         arguments.append(args) 
   ...:         return return_value 
   ...:     return function 
   ...:                                                                         

In [3]: mock_return(3)                                                          
Out[3]: <function __main__.mock_return.<locals>.function(*args, **kwargs)>

In [4]: a = mock_return(3)                                                      

In [5]: a(4, 5, 6)                                                              
Out[5]: 3

In [6]: arguments                                                               
Out[6]: [(4, 5, 6)]
```

pauljxtan replied:
Yes, any PATCH request with any combination of arguments will return the same response.

That's a good point about validating the arguments. I'll look at adding that into the relevant tests :+1:

pauljxtan replied:
I've added an argument check to the test - I'm especially grateful that you suggested this because it actually caught a bug!

antalszava commented on tests/frontend/test_engine.py:
```diff
@@ -166,209 +175,299 @@ def inspect():
-        monkeypatch.setattr(starship_engine, "_queue_job", lambda job_content: mock_job)
+    def test_get_job_result(self, connection, monkeypatch):
+        """Tests a successful job result request."""
+        result_samples = np.array([[1, 2], [3, 4]], dtype=np.int8)
```
The `dtype` argument could be parametrized as well, just to see that it works fine for types like `np.float64` and the likes as well :)

pauljxtan replied:
Good idea, will do so!

pauljxtan replied:
This has been added!

antalszava commented on tests/frontend/test_engine.py:
```diff
@@ -166,209 +175,299 @@ def inspect():
+                requests,
+                "get",
+                mock_return(mock_response(200, binary_body=buf.getvalue())),
+            )
```
Cool! :)

antalszava commented on tests/frontend/test_engine.py:
```diff
@@ -166,209 +175,299 @@ def inspect():
+        connection.cancel_job("123")
+
+    def test_cancel_job_error(self, connection, monkeypatch):
+        """Tests a successful job cancellation request."""
```
```suggestion
        """Tests a failed job cancellation request."""
```

antalszava commented on tests/frontend/test_engine.py:
```diff
@@ -166,209 +174,304 @@ def inspect():
+        for _ in range(server.REQUESTS_BEFORE_COMPLETED):
+            job.refresh()
+
+        assert job.status == JobStatus.COMPLETED
```
Nice!

josh146 commented on strawberryfields/engine.py:
```diff
@@ -531,161 +544,507 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
+        # (After some time...)
+        job.refresh()
+        job.status  # <JobStatus.COMPLETED: 'complete'>
+        job.result  # [[0 1 0 2 1 0 0 0]]
```
This might be easier to read using Python prompt syntax. Eg:

```rest
Run a job synchronously:

>>> engine = StarshipEngine("chip2")
>>> result = engine.run(program, shots=1) # blocking call
>>> result
[[0 1 0 2 1 0 0 0]]

Run a job synchronously, but cancel it before it is completed
using a keyboard interrupt (`ctrl+c`):

>>> result = engine.run(program, shots=1)
^C---------------------------------------------------------------------------
KeyboardInterrupt                         Traceback (most recent call last)
<ipython-input-4-f1a1495c6d9c> in <module>()
----> 1 time.sleep(10)

Run a job asynchronously:

>>> job = engine.run_async(program, shots=1)
>>> job.status 
<JobStatus.QUEUED: 'queued'>
>>> job.result
InvalidJobOperationError
>>> job.refresh()
>>> job.status 
<JobStatus.COMPLETED: 'complete'>
>>> job.result
[[0 1 0 2 1 0 0 0]]
```

The nice thing about using prompts is that Sphinx renders it automatically without needing to specify a code-block or indent

pauljxtan replied:
Oh nice! I'll update all the relevant docstrings to use the prompt syntax. (I assume these won't be run as doctests?)

pauljxtan replied:
The docstrings for `StarshipEngine` and `Connection` have been updated to use the prompt syntax.

josh146 commented on strawberryfields/engine.py:
```diff
@@ -531,161 +544,507 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
+    def run_async(self, program: Program, shots: int = 1) -> Job:
+        """Runs a remote job asynchronously.
+
+        In the asynchronous mode, a `Job` is returned immediately, and the user can
```
```suggestion
        In the asynchronous mode, a ``Job`` is returned immediately, and the user can
```

josh146 commented on tests/frontend/test_engine.py:
```diff
@@ -166,209 +174,304 @@ def inspect():
-    """
-    Tests various methods on the remote engine StarshipEngine.
+@pytest.fixture
+def connection():
```
All API specific tests I would move into their own folder (i.e., `tests/api/test_connection.py`, `tests/api/test_job.py`)

josh146 replied:
It might even make sense to move the API-specific starship tests here as well.

pauljxtan replied:
Sounds good, I will do so!

pauljxtan replied:
All of the new tests have been moved under a new `api` directory.

josh146 commented on tests/frontend/test_engine.py:
```diff
@@ -166,209 +174,304 @@ def inspect():
-        inputs = MagicMock()
+        result = Result([[1, 2], [3, 4]], is_stateful=False)
+
+        with pytest.raises(AttributeError):
```
It would be good to also include the `match=` argument here, just to make sure that the _right_ attribute error is raised!

pauljxtan replied:
Thanks, I've added `match` checks to all the relevant tests.

josh146 commented on tests/frontend/test_engine.py:
```diff
@@ -166,209 +174,304 @@ def inspect():
+        """Tests that `job.result` raises an error for an incomplete job."""
+        job = Job("abc", status=JobStatus.QUEUED, connection=Connection)
+
+        with pytest.raises(AttributeError):
```
Same applies throughout (better to match the error message in these assertions)

pauljxtan replied:
Done :+1:

pauljxtan commented on starship:
```diff
@@ -14,114 +14,51 @@
+
+    eng = StarshipEngine("chip2", connection)
+    sys.stdout.write("Computing...\n")
+    result = eng.run(program, shots=1)
```
Ah yes, I think you're right. I'll remove that, thanks!

pauljxtan commented on strawberryfields/engine.py:
```diff
@@ -531,161 +544,541 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
+    """Represents the status of a remote job.
+
+    This class maps a set of job statuses to the string representations returned by the
+    remote platform.
```
I think the first option sounds good, just to separate things more cleanly and have a dedicated namespace. Which do you prefer?

pauljxtan replied:
Oh, never mind, I see you said it in another comment :smiley:

pauljxtan commented on strawberryfields/engine.py:
```diff
@@ -531,161 +544,541 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
+        host: str = "platform.strawberryfields.ai",
+        port: int = 443,
+        use_ssl: bool = True,
+    ):
```
Done!

co9olguy commented on strawberryfields/api/connection.py:
```diff
@@ -0,0 +1,284 @@
+    >>> job
+    <Job: id=d177cbf5-1816-4779-802f-cef2c117dc1a, ...>
+    >>> job.status
+    <JobStatus.QUEUED: 'queued'>
```
Users won't likely ever use a `Connection` directly, but they will certainly interact with a `Job`. The string printed here could be improved to be simpler and less redundant, e.g., `<JobStatus: queued>`

josh146 replied:
Another thought - if the user wants to programmatically check job status, they will need to import the `JobStatus` enum from within the API. Will they also be able to compare via a simple string? i.e., `job.status == "queued"`?

pauljxtan replied:
They can compare to a string with `job.status.value == "queued"`. It's not ideal, but since the alternative is dropping the enum entirely, I feel that it's a decent tradeoff :thinking: 

I guess the options here are:
- Leave it as is
- Override the `__eq__` method to enable comparison with a string, i.e. `JobStatus.QUEUED == "queued"` -> `True`
- Add properties on `JobStatus`, e.g. `is_queued`, `is_completed`, etc.
- Get rid of the enum and replace all usages with string literals

What do you think?

pauljxtan replied:
Actually, please ignore the above.

After more thought, the obvious solution (to me) would be to keep `JobStatus` internally but expose it as a string via the property. I will make that change - let me know if you think there is a better way!

co9olguy commented on strawberryfields/api/connection.py:
```diff
@@ -0,0 +1,284 @@
+    >>> job.status
+    <JobStatus.QUEUED: 'queued'>
+    >>> job.result
+    AttributeError
```
Is this the exact string that will be printed? Honestly not sure what the best way is to "print" an exception in a docstring :thinking:

pauljxtan replied:
The exact output has a bunch of traceback info preceding it. It seemed a little distracting/superfluous so I omitted it, but I can add that in if we consider it necessary.

pauljxtan replied:
Nice, I've updated the docstring to use something similar (with `AttributeError` and the attached error message).

co9olguy commented on strawberryfields/api/connection.py:
```diff
@@ -0,0 +1,284 @@
+    >>> job
+    <Job: id=59a1c0b1-c6a7-4f9b-ae37-0ac5eec9c413, ...>
+    >>> job.status
+    <JobStatus.COMPLETED: 'complete'>
```
see above comment

pauljxtan replied:
This one is the actual string as implemented by `__str__`/`__repr__`.

co9olguy replied:
oh, for this I was referring to the comment about `<JobStatus.QUEUED: 'queued'>`

co9olguy commented on strawberryfields/api/connection.py:
```diff
@@ -0,0 +1,284 @@
+        """The API authentication token.
+
+        Returns:
+            str: the authentication token
```
Docstring seems redundant. See my general question (may appear elsewhere in PR) about whether it is even necessary to report the `Return` of a `@property`

co9olguy commented on strawberryfields/api/connection.py:
```diff
@@ -0,0 +1,284 @@
+        path = "/jobs/{}".format(job_id)
+        response = requests.get(self._url(path), headers=self._headers)
+        if response.status_code == 200:
+            return Job(
```
Might be worthwhile to make this `Job`(...)` block a separate function, since it is used elsewhere in the same way

pauljxtan replied:
Do you mean the `Job(...)` "constructor"? Since it is already is (sort of) a function in itself, I'm not sure I see the benefit of factoring it out further, unless I'm missing something :thinking:

co9olguy commented on strawberryfields/api/connection.py:
```diff
@@ -0,0 +1,284 @@
+            "Failed to get job result: {}".format(self._format_error_message(response))
+        )
+
+    def cancel_job(self, job_id: str):
```
Better to use `id_` as used elsewhere?

pauljxtan replied:
I think using `id_` within the `Job` class itself makes sense, since it is quite obvious what the variable is identifying; but I think `job_id` is more explicit for methods on other classes and makes the interface a bit clearer.

pauljxtan commented on tests/api/test_starship_engine.py:
```diff
@@ -0,0 +1,97 @@
+
+@pytest.fixture
+def job_to_complete(connection, monkeypatch):
+    """Mocks a remote job that is completed after a certain number of requests."""
```
It's not very clear and I'm not enamoured with how the code looks, but basically the `MockServer` (instantiated as `server` here) holds the number of requests as a state attribute. It returns a job status of "queued" if the number of requests is below a certain number, then returns "completed" once it exceeds the threshold.

It's a little hacky but I think it does the job in lieu of a full integration test (which we plan to tackle in a future PR).

pauljxtan commented on tests/api/test_starship_engine.py:
```diff
@@ -0,0 +1,97 @@
+        mock_return(Job(id_="123", status=JobStatus.OPEN, connection=connection)),
+    )
+    server = MockServer()
+    monkeypatch.setattr(Connection, "get_job_status", server.get_job_status)
```
I agree that it looks a little jarring, but the files in this PR are all auto-formatted with `black` (which has a default line length limit of 88 characters), so I'm hesitant to manually impose ad hoc formatting :slightly_smiling_face:

josh146 replied:
@pauljxtan, we actually use a line length of 100 characters on SF! So you might need to reblack before merging (`black -l 100 strawberryfields/`).

Before the next release, I am going to add a CI check that ensures blacking is done with the right line length, sorry about that!

pauljxtan replied:
Ah I see, I will reformat the relevant files - thanks for pointing that out!

pauljxtan replied:
@josh146 - running `black -l 100` it on the full `strawberryfields/` seemed to change a lot of unrelated modules, so I've run it only on the files relevant to this PR.

co9olguy replied:
good idea @pauljxtan

josh146 replied:
:+1: Just prior to the new SF release, I plan on blacking the entire library, and adding a formatting CI check. This will stop new PRs from allowing unformatted code through.

pauljxtan commented on tests/api/test_starship_engine.py:
```diff
@@ -0,0 +1,97 @@
+
+    def test_run_complete(self, connection, prog, job_to_complete):
+        """Tests a successful synchronous job execution."""
+        engine = StarshipEngine("chip2", connection=connection)
```
Ah, thanks for letting me know! I'll make a note to change this once the new names are finalized.

pauljxtan commented on tests/api/test_starship_engine.py:
```diff
@@ -0,0 +1,97 @@
+        engine = StarshipEngine("chip2", connection=connection)
+        result = engine.run(prog)
+
+        assert np.array_equal(result.samples.T, np.array([[1, 2], [3, 4]]))
```
Please see: https://github.com/XanaduAI/strawberryfields/pull/294#discussion_r383527915

Long story short, @thisac has already resolved this on `master` but as this branch is based on `demo`, I felt that it seemed less of a headache to address in a future commit (probably when these changes get merged into `master`) as various tests throughout the codebase would have to updated.

josh146 replied:
@pauljxtan I've been periodically merging master into demo every week, to make sure they keep in sync and lesson merge conflicts. The changes made by @thisac should already be in the demo branch now, if you want to merge them in here.

pauljxtan replied:
Thanks for letting me know! I've merged the latest `demo` and made a small change to `Result.__init__` removing the `np.stack` (which is no longer needed). I've modified the new API tests accordingly (removed all transposes), and all relevant existing tests also appear to be passing.

pauljxtan commented on strawberryfields/engine.py:
```diff
@@ -19,120 +19,21 @@
-        return "Result: {} subsystems, state: {}\n samples: {}".format(
-            len(self.samples), self.state, self.samples
-        )
+log = logging.getLogger(__name__)
```
That was used to log a warning message in `StarshipEngine.run()`.

pauljxtan commented on tests/api/test_connection.py:
```diff
@@ -0,0 +1,215 @@
+        with pytest.raises(RequestFailedError, match="Failed to create job"):
+            connection.create_job("chip2", prog, 1)
+
+    @pytest.mark.skip(reason="method not yet implemented")
```
Thanks, updated.

pauljxtan commented on tests/api/test_connection.py:
```diff
@@ -0,0 +1,215 @@
+
+        assert [job.id for job in jobs] == [str(i) for i in range(5, 10)]
+
+    @pytest.mark.skip(reason="method not yet implemented")
```
And updated.

pauljxtan commented on tests/api/test_starship_engine.py:
```diff
@@ -0,0 +1,97 @@
+from strawberryfields.engine import StarshipEngine
+
+
+def mock_return(return_value):
```
Good catch, updated.

pauljxtan commented on tests/api/test_job.py:
```diff
@@ -0,0 +1,41 @@
+
+    def test_final_job_raises_on_cancel(self, connection):
+        """Tests that `job.cancel()` raises an error for a complete, failed, or
+        aleady cancelled job."""
```
Thanks for pointing that out, updated.

pauljxtan commented on strawberryfields/engine.py:
```diff
@@ -531,161 +441,151 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
-    def __init__(self, backend, polling_delay_seconds=1, **kwargs):
-        super().__init__(backend)
+    The following examples instantiate an engine with the default configuration, and
+    runs jobs both synchronously and asynchronously.
```
Not sure :thinking: I like the symmetry of "synchronous" as the opposite of "asynchronous", but open to changing it if we don't find that clear enough!

josh146 replied:
Perhaps 'blocking' and 'non-blocking' might make more sense to people coming from a scientific computing background?

pauljxtan replied:
Sounds good! I've updated the relevant parts to use those terms.

pauljxtan commented on strawberryfields/engine.py:
```diff
@@ -531,161 +441,151 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
-        """
-        # Do nothing for now...
-        pass
+    Run a job synchronously, but cancel it before it is completed using a keyboard
```
Good point. I've removed the example keyboard interrupt and added a line to the docstring for `run()` describing the feature.

pauljxtan commented on tests/api/test_connection.py:
```diff
@@ -0,0 +1,215 @@
+        """Tests a failed ping to the remote host."""
+        monkeypatch.setattr(requests, "get", mock_return(MockResponse(500, {})))
+
+        assert not connection.ping()
```
:stuck_out_tongue:

pauljxtan commented on strawberryfields/engine.py:
```diff
@@ -531,161 +441,151 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
 
         Returns:
-            (strawberryfields.api_client.Job): a Job instance referencing the queued job
+            str: the target device used by the engine
```
Good point - I've removed the "Returns" part from the relevant property docstrings.

pauljxtan replied:
On second thought, we probably want at least the types in there. I've put them back in, but without the repetitive text.

pauljxtan commented on strawberryfields/engine.py:
```diff
@@ -531,161 +441,151 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
+        """The connection object used by the engine.
+
+        Returns:
+            strawberryfields.engine.Connection: the connection object used by the engine
```
Please see above.

pauljxtan commented on strawberryfields/api/job.py:
```diff
@@ -0,0 +1,139 @@
+        Only an open or queued job can be refreshed; an exception is raised otherwise.
+        """
+        if self.status.is_final:
+            log.warning("A %s job cannot be refreshed", self.status.value)
```
I believe it's the recommended way of formatting `logging` messages.

https://stackoverflow.com/questions/34619790/pylint-message-logging-format-interpolation

co9olguy replied:
Didn't know, thanks

josh146 commented on tests/api/conftest.py:
```diff
@@ -0,0 +1,44 @@
+    """A helper function for defining a mock function that returns the given value for
+    any arguments.
+    """
+    return lambda *args, **kwargs: return_value
```
Having separate conftests makes it much clearer which fixtures are used in which part of the test suite!

pauljxtan commented on starship:
```diff
@@ -14,114 +14,51 @@
     program = load(args.input)
-    eng = StarshipEngine(program.target)
+
+    eng = StarshipEngine("chip2", connection)
```
Good catch - I've changed that to use `program.target` as before.

pauljxtan commented:
Hey @antalszava @co9olguy @josh146, I think I've addressed all of your comments, but let me know if anything is missing. (GitHub seems to be having some issues at the moment causing comments to be intermittently dropped.)

antalszava commented on doc/index.rst:
```diff
@@ -174,7 +174,6 @@ Strawberry Fields is **free** and **open source**, released under the Apache Lic
    code/sf_apps
    code/sf_ops
    code/sf_utils
-   code/sf_api_client
```
Instead, the new `api` directory will have to be included, such that when `make docs` is run, the new directory is documented.

pauljxtan replied:
I added a `sf_api.rst` as you described, and I _think_ the docs are showing up properly in the expected places - but would appreciate if someone could confirm I did it right!

antalszava replied:
@pauljxtan thanks so much! :blush: Could you move the line with the `api` change here? (such that it is in alphabetical order, this is also the order in which it appears on the website)

antalszava commented on strawberryfields/api/job.py:
```diff
@@ -0,0 +1,139 @@
+        Returns:
+            bool: ``True`` if the job status is final, and ``False`` otherwise
+        """
+        return self in (JobStatus.CANCELLED, JobStatus.COMPLETED, JobStatus.FAILED)
```
(not too sure if there was a change in the meantime, but cannot really see the diff betteen the code and the suggestion)

co9olguy replied:
yep, ignore this for now. I was reviewing at the same time that @pauljxtan was updating with new commits

antalszava commented on strawberryfields/api/result.py:
```diff
@@ -0,0 +1,116 @@
+    def samples(self):
+        """Measurement samples.
+
+        Returned measurement samples will have shape ``(modes,)``. If multiple
```
```suggestion
        Returned measurement samples will have shape ``(shots, modes)``. If multiple
```

antalszava commented on strawberryfields/api/result.py:
```diff
@@ -0,0 +1,116 @@
+
+        Returned measurement samples will have shape ``(modes,)``. If multiple
+        shots are requested during execution, the returned measurement samples
+        will instead have shape ``(shots, modes)``.
```
```suggestion
        Returned measurement samples will have shape ``(shots, modes)``.
```

antalszava replied:
As per #300

pauljxtan replied:
Good catch :+1:

antalszava commented on strawberryfields/api/result.py:
```diff
@@ -0,0 +1,115 @@
+
+    def __str__(self):
+        """String representation."""
+        return "Result: {} subsystems, state: {}\n samples: {}".format(
```
Hmm, perhaps only smaller states could be outputted by this and there could be a separate auxiliary method that outputs every state.

That is, this could have a check along the lines of
```python
if len(self.state)>10 or len(self.samples)>1000:
# output lengths and advice users to use the Result.verbose_str option
# else proceed with outputting everything
```
Open to suggestions if this could work and if so, what the limit for these should be.

co9olguy replied:
Might be possible that the string-printing methods of numpy already take the size into account and truncate (haven't tested myself)

pauljxtan replied:
Sorry, could've sworn I replied to this but maybe the GitHub outage ate it.

I tested this out, and it does seem that the samples get automatically truncated:

```python
In [4]: result = Result(np.random.randint((1e6, 8)))

In [6]: str(result)
Out[6]: 'Result: 1 subsystems, state: None\n samples: [[148373      4]]'
```

I should also note that this was a holdover from existing code that I simply moved to a different module, so I assume this was probably also discussed when it was first implemented?

pauljxtan replied:
Oh, my bad, got confused with the syntax there :stuck_out_tongue: Thanks for checking that!

thisac replied:
@josh146 It seems like it's the `samples = np.stack(samples, 1)` that is taking time when printing the results (I'm using the master branch). This should be gone in this PR though, since it's an unnecessary line and apparently quite slow (which isn't too surprising since it not only transposes the array but also re-stacks all the inputs; I'm not sure why that was done that way though).

josh146 replied:
Nice, good to know :+1:

antalszava commented on strawberryfields/engine.py:
```diff
@@ -552,161 +467,141 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
-    """
-    Starship quantum program executor engine.
+class StarshipEngine:
+    """A quantum program executor engine that that provides a simple interface for
```
```suggestion
    """A quantum program executor engine that provides a simple interface for
```

co9olguy commented on tests/api/test_job.py:
```diff
@@ -32,14 +34,31 @@ def test_incomplete_job_raises_on_result_access(self, connection):
             match="The result is undefined for jobs that are not completed",
         ):
-            _ = job.result
+            job.result
```
@pauljxtan I see, was the linter complaining if you didn't assign the property to something? In that case, I might defer to the linter (sorry for making pointless work :laughing:)

pauljxtan replied:
Haha yes it was :slightly_smiling_face:

co9olguy commented on strawberryfields/engine.py:
```diff
@@ -297,7 +299,7 @@ def _broadcast_nones(val, dim):
 
         if self.samples is not None:
-            return Result(self.samples.copy())
+            return Result(np.array(self.samples).T)
```
@pauljxtan @thisac How come there is a transpose needed here? Is there still an inconsistent return shape between some parts of the pipeline?

pauljxtan replied:
I think it's due to the way the samples are computed by the simulator backend (?) I don't understand the local backend well enough to modify the code there without a risk of breaking something, so will have to defer to someone else's expertise here :slightly_smiling_face:

thisac replied:
`self.samples` has shape `(modes, shots)` and thus needs to be transposed to `(shots, modes)` somewhere. In the master branch the transposition happens in the `Results` class at the line `np.stack(samples, 1)`, so the `BaseEngine._run` function simply returns `self.samples.copy()` (with shape `(modes, shots)`). I'm not sure why this is. It was like that before I touched it; I only added the specific check in `Results` for a single-shot list so that it returned an array of shape `(shots, modes)` even in that case. But this was all done in the master branch, and not the demo branch. Now both branches seem to have different implementations of the `Results` class. I'm not too sure about what the demo branch is though, so maybe this doesn't matter that much. :slightly_smiling_face:

pauljxtan replied:
Thanks! This is making more sense to me now. What I essentially did was move the transposition slightly down the pipeline, from `Result.__init__` to the end of `BaseEngine._run`. This prevents `np.stack(samples, 1)` (now removed) from incorrectly transposing samples coming in from the remote platform (which have `(shots, modes)`). So I believe this all works out.

thisac replied:
That makes sense. Thanks! :slightly_smiling_face:  Should this change preferably be done in the master branch as well (or will the demo branch be merged with the master branch at some point)? @josh146 @co9olguy

josh146 replied:
The plan is to merge the demo branch into master in the coming week(s), before the upcoming SF release!

josh146 replied:
However, it sounds like there will be a merge conflict now, as master does the transposition in `Result.__init__`, and demo does it in `BaseEngine._run`, so we will need to be careful to resolve this conflict when we merge.

josh146 replied:
I think we will have to look at this again carefully after the merge. In an ideal world, the API _as well as_ the Gaussian/Fock/TF backends, should be returning samples of the form `(shots, modes)`. The engine or the results class should not have to do any transposition.

thisac replied:
I looked through how the `LocalEngine` gets samples from the backends, and it doesn't look like there's an easy fix to get the correct shape directly from the backends. The only backend that supports multi-shot measurements seems to be the Gaussian backend, and it returns an array of values for one mode at a time (one value for each shot). In that case the combining of modes is done in `engine.py` either way, so the transposing would have to be there also. Maybe I misunderstood something though.

As for the other backends they don't support multiple shots together with any measurements, and simply return a list of length = number of modes filled with `None`, which is then converted into a nested array in the desired shape in `engine.py` (the `_broadcast_nones` function in `BaseEngine._run`).

co9olguy commented on strawberryfields/engine.py:
```diff
@@ -443,45 +467,32 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
 
     The following examples instantiate an engine with the default configuration, and
-    runs jobs both synchronously and asynchronously.
+    runs both blocking and non-blocking jobs.
```
```suggestion
    run both blocking and non-blocking jobs.
```

antalszava commented on doc/index.rst:
```diff
@@ -185,3 +184,4 @@ Strawberry Fields is **free** and **open source**, released under the Apache Lic
    code/sf_program
    code/sf_program_utils
    code/sf_parameters
+   code/sf_api
```
So moving this line for where the deleted line is

pauljxtan replied:
Thanks for pointing that out, done :+1:

pauljxtan commented on starship:
```diff
@@ -14,114 +14,52 @@
     group.add_argument("--input", "-i", help="the xbb file to run")
+    group.add_argument(
+        "--ping", "-p", action="store_true", help="test the API connection"
+    )
```
Hmm, after taking another look at this, it occurs to me that the API token should probably be loaded in the same way as the other connection parameters (host, port, etc.) via the new configuration logic implemented by Antal. This is more consistent, would reduce the required arguments down to the bare minimum (input file), and we wouldn't even need to instantiate a `Connection` object in the CLI script - allowing us to work at the highest possible level of abstraction with only an `Engine`.

I've removed the `--token` parameter from the CLI - which can now be run simply as:

```
$ ./starship -i test.xbb
Executing program on remote hardware...
[[0 0 1 1 0 0 1 0]]%
```

antalszava replied:
No worries @pauljxtan, if further final touches are required to the `starship` CLI, they could come down the line in upcoming separate PRs. :blush:

pauljxtan commented on doc/code/sf_api.rst:
```diff
@@ -0,0 +1,14 @@
+.. automodapi:: strawberryfields.api
+    :no-heading:
+    :no-inheritance-diagram:
+    :include-all-objects:
```
Yes, @antalszava has been a massive help on various things :+1:

pauljxtan commented:
Planning to merge this afternoon if/when the latest CI build passes.

thisac commented on strawberryfields/api/result.py:
```diff
@@ -0,0 +1,114 @@
+        state: <GaussianState: num_modes=3, pure=True, hbar=2>
+        samples: [0, 0, 0]
+    >>> results.samples
+    [0, 0, 0]
```
```suggestion
        samples: [[0, 0, 0]]
    >>> results.samples
    np.array([[0, 0, 0]])
```

Sorry for the late suggestion. Just noticed this.

pauljxtan replied:
Good catch, thanks!

thisac commented:
I've looked through the files relevant to the changes I made for the unified results PR earlier, and it looks good. This is a much better solution and I don't think that anything will break when merging (at least not anything obvious, related to the `Result` class as well as the transposition being relocated). :+1:

pauljxtan commented:
> I've looked through the files relevant to the changes I made for the unified results PR earlier, and it looks good. This is a much better solution and I don't think that anything will break when merging (at least not anything obvious, related to the `Result` class as well as the transposition being relocated). +1

Thanks for taking the time to look this over @thisac!
